File: benchmarks/80-shgo-testbed/shgo_testbed.m
(* Experiment 80 -- SHGO across the NMinimize test corpus (NMinimize,
   Method -> {"SHGO", ...}).

   This runs the *prior NMinimize stress/unit test problems* (from
   tests/test_nminimize.c) through the new SHGO engine and races each against
   scipy.optimize.shgo on the identical problem (same box, sampling method, and
   sample count). It is the cleanly-comparable set -- the multimodal stress
   functions and the inequality-constrained problems within SHGO's applicability
   envelope, where both implementations reach the same global and the timing
   comparison is meaningful. The rest of the corpus is analysed in README.md:
   the constrained problems that scipy.optimize.shgo mishandles (Mathilda solves
   them via its augmented-Lagrangian polish), the high-dimensional / sharp-needle
   functions outside SHGO's envelope (both implementations miss the global), and
   the combinatorial-integer problems (scipy.optimize.shgo is continuous-only).

   The check is the objective at the global optimum, rounded per problem so that
   both implementations, reaching the same basin, agree. Indexed objectives use
   x[i] exactly as the original tests do. *)

Get["../harness.m"];

require[{"NMinimize"}];

(* ================= Multimodal battery (low-dimensional) ================= *)

bench["01 Eggholder 2D", NMinimize[{-(y+47)Sin[Sqrt[Abs[y+x/2+47]]]-x Sin[Sqrt[Abs[x-(y+47)]]], -512<=x<=512 && -512<=y<=512}, {x,y}, Method->{"SHGO","SearchPoints"->500}];];
check["01 Eggholder 2D", Round[First[NMinimize[{-(y+47)Sin[Sqrt[Abs[y+x/2+47]]]-x Sin[Sqrt[Abs[x-(y+47)]]], -512<=x<=512 && -512<=y<=512}, {x,y}, Method->{"SHGO","SearchPoints"->500}]]]];

bench["02 Schwefel 2D", NMinimize[{837.9658-x Sin[Sqrt[Abs[x]]]-y Sin[Sqrt[Abs[y]]], -500<=x<=500 && -500<=y<=500}, {x,y}, Method->{"SHGO","SearchPoints"->300}];];
check["02 Schwefel 2D", Round[10^3 First[NMinimize[{837.9658-x Sin[Sqrt[Abs[x]]]-y Sin[Sqrt[Abs[y]]], -500<=x<=500 && -500<=y<=500}, {x,y}, Method->{"SHGO","SearchPoints"->300}]]]];

bench["03 Schaffer N2 2D", NMinimize[{0.5+(Sin[x^2-y^2]^2-0.5)/(1+0.001(x^2+y^2))^2, -100<=x<=100 && -100<=y<=100}, {x,y}, Method->{"SHGO","SearchPoints"->400}];];
check["03 Schaffer N2 2D", Round[10^4 First[NMinimize[{0.5+(Sin[x^2-y^2]^2-0.5)/(1+0.001(x^2+y^2))^2, -100<=x<=100 && -100<=y<=100}, {x,y}, Method->{"SHGO","SearchPoints"->400}]]]];

bench["04 Mishra Bird disk", NMinimize[{Sin[y]Exp[(1-Cos[x])^2]+Cos[x]Exp[(1-Sin[y])^2]+(x-y)^2, (x+5)^2+(y+5)^2<25}, {x,y}, Method->{"SHGO","SearchPoints"->300}];];
check["04 Mishra Bird disk", Round[10^3 First[NMinimize[{Sin[y]Exp[(1-Cos[x])^2]+Cos[x]Exp[(1-Sin[y])^2]+(x-y)^2, (x+5)^2+(y+5)^2<25}, {x,y}, Method->{"SHGO","SearchPoints"->300}]]]];

bench["05 Rugged sine 2D", NMinimize[{x^2+y^2+10 Sin[3x]^2+10 Sin[3y]^2, -5<=x<=5 && -5<=y<=5}, {x,y}, Method->{"SHGO","SearchPoints"->200}];];
check["05 Rugged sine 2D", Round[10^4 First[NMinimize[{x^2+y^2+10 Sin[3x]^2+10 Sin[3y]^2, -5<=x<=5 && -5<=y<=5}, {x,y}, Method->{"SHGO","SearchPoints"->200}]]]];

bench["06 Ackley 3D", NMinimize[{-20 Exp[-0.2 Sqrt[1/3 Sum[x[i]^2,{i,1,3}]]]-Exp[1/3 Sum[Cos[2 Pi x[i]],{i,1,3}]]+20+E, Table[-32<=x[i]<=32,{i,1,3}]}, Table[x[i],{i,1,3}], Method->{"SHGO","SamplingMethod"->"Simplicial","SearchPoints"->200}];];
check["06 Ackley 3D", Round[10^4 First[NMinimize[{-20 Exp[-0.2 Sqrt[1/3 Sum[x[i]^2,{i,1,3}]]]-Exp[1/3 Sum[Cos[2 Pi x[i]],{i,1,3}]]+20+E, Table[-32<=x[i]<=32,{i,1,3}]}, Table[x[i],{i,1,3}], Method->{"SHGO","SamplingMethod"->"Simplicial","SearchPoints"->200}]]]];

bench["07 Rastrigin 5D", NMinimize[{50+Sum[x[i]^2-10 Cos[2 Pi x[i]],{i,1,5}], Table[-5.12<=x[i]<=5.12,{i,1,5}]}, Table[x[i],{i,1,5}], Method->{"SHGO","SamplingMethod"->"Simplicial","SearchPoints"->300}];];
check["07 Rastrigin 5D", Round[10^4 First[NMinimize[{50+Sum[x[i]^2-10 Cos[2 Pi x[i]],{i,1,5}], Table[-5.12<=x[i]<=5.12,{i,1,5}]}, Table[x[i],{i,1,5}], Method->{"SHGO","SamplingMethod"->"Simplicial","SearchPoints"->300}]]]];

bench["08 Griewank 5D", NMinimize[{Sum[x[i]^2/4000,{i,1,5}]-Product[Cos[x[i]/Sqrt[i]],{i,1,5}]+1, Table[-15<=x[i]<=15,{i,1,5}]}, Table[x[i],{i,1,5}], Method->{"SHGO","SamplingMethod"->"Simplicial","SearchPoints"->300}];];
check["08 Griewank 5D", Round[10^4 First[NMinimize[{Sum[x[i]^2/4000,{i,1,5}]-Product[Cos[x[i]/Sqrt[i]],{i,1,5}]+1, Table[-15<=x[i]<=15,{i,1,5}]}, Table[x[i],{i,1,5}], Method->{"SHGO","SamplingMethod"->"Simplicial","SearchPoints"->300}]]]];

(* ================= Inequality-constrained ================= *)

bench["09 Disk linear", NMinimize[{x+y, x^2+y^2<=9}, {x,y}, Method->{"SHGO","SearchPoints"->100}];];
check["09 Disk linear", Round[10^4 First[NMinimize[{x+y, x^2+y^2<=9}, {x,y}, Method->{"SHGO","SearchPoints"->100}]]]];

bench["10 Chained trig 1D", NMinimize[{Sin[2x]+Cos[x], -2<=x<=3}, {x}, Method->{"SHGO","SearchPoints"->100}];];
check["10 Chained trig 1D", Round[10^4 First[NMinimize[{Sin[2x]+Cos[x], -2<=x<=3}, {x}, Method->{"SHGO","SearchPoints"->100}]]]];

bench["11 Quadratic halfplane", NMinimize[{x^2+y^2, x+y>=1}, {x,y}, Method->{"SHGO","SearchPoints"->100}];];
check["11 Quadratic halfplane", Round[10^4 First[NMinimize[{x^2+y^2, x+y>=1}, {x,y}, Method->{"SHGO","SearchPoints"->100}]]]];

File: benchmarks/80-shgo-testbed/shgo_testbed.py
#!/usr/bin/env python3
"""Experiment 80 -- SHGO across the NMinimize test corpus (scipy column).

Same 11 problems as ``shgo_testbed.m``, same labels, same order, solved with
scipy.optimize.shgo using the SAME sampling_method and sample count (n) and an
equivalent bounded box + constraints, so the race is SHGO-vs-shgo. This is the
cleanly-comparable set (the multimodal stress functions and the
inequality-constrained problems). The rest of the corpus is analysed in
README.md -- the constrained problems scipy.optimize.shgo mishandles, the
high-dimensional / sharp-needle functions outside SHGO's envelope, and the
combinatorial-integer problems (scipy.optimize.shgo is continuous-only).

The check is the objective at the global optimum, rounded per problem so that
both implementations, reaching the same basin, agree.
"""

import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import numpy as np
from scipy.optimize import shgo

from harness import bench, check, require

require(["scipy.optimize:shgo"])


def _solve(fun, bounds, n, method, constraints=()):
    return shgo(fun, bounds, n=n, iters=1, sampling_method=method,
                constraints=constraints)


def _val(res, scale):
    f = res.fun
    if f is None or not np.isfinite(f):
        return "NOFEASIBLE"
    return int(np.floor(scale * f + 0.5))


def _eggholder(v):
    x, y = v
    return (-(y + 47) * np.sin(np.sqrt(abs(y + x / 2 + 47)))
            - x * np.sin(np.sqrt(abs(x - (y + 47)))))

def _schwefel2(v):
    x, y = v
    return 837.9658 - x * np.sin(np.sqrt(abs(x))) - y * np.sin(np.sqrt(abs(y)))

def _schaffer2(v):
    x, y = v
    return 0.5 + (np.sin(x * x - y * y) ** 2 - 0.5) / (1 + 0.001 * (x * x + y * y)) ** 2

def _mishra(v):
    x, y = v
    return (np.sin(y) * np.exp((1 - np.cos(x)) ** 2)
            + np.cos(x) * np.exp((1 - np.sin(y)) ** 2) + (x - y) ** 2)

def _rugged(v):
    x, y = v
    return x * x + y * y + 10 * np.sin(3 * x) ** 2 + 10 * np.sin(3 * y) ** 2

def _ackley3(v):
    v = np.asarray(v)
    return (-20 * np.exp(-0.2 * np.sqrt(np.sum(v * v) / 3))
            - np.exp(np.sum(np.cos(2 * np.pi * v)) / 3) + 20 + np.e)

def _rastrigin5(v):
    v = np.asarray(v)
    return 50 + float(np.sum(v * v - 10 * np.cos(2 * np.pi * v)))

def _griewank5(v):
    v = np.asarray(v)
    idx = np.arange(1, 6)
    return float(np.sum(v * v) / 4000 - np.prod(np.cos(v / np.sqrt(idx))) + 1)


_c01 = lambda: _solve(_eggholder,  [(-512, 512)] * 2,   500, "simplicial")
_c02 = lambda: _solve(_schwefel2,  [(-500, 500)] * 2,   300, "simplicial")
_c03 = lambda: _solve(_schaffer2,  [(-100, 100)] * 2,   400, "simplicial")
_c04 = lambda: _solve(_mishra,     [(-10, 10)] * 2,     300, "simplicial",
                      ({"type": "ineq", "fun": lambda v: 25 - (v[0] + 5) ** 2 - (v[1] + 5) ** 2},))
_c05 = lambda: _solve(_rugged,     [(-5, 5)] * 2,       200, "simplicial")
_c06 = lambda: _solve(_ackley3,    [(-32, 32)] * 3,     200, "simplicial")
_c07 = lambda: _solve(_rastrigin5, [(-5.12, 5.12)] * 5, 300, "simplicial")
_c08 = lambda: _solve(_griewank5,  [(-15, 15)] * 5,     300, "simplicial")
_c09 = lambda: _solve(lambda v: v[0] + v[1], [(-10, 10)] * 2, 100, "simplicial",
                      ({"type": "ineq", "fun": lambda v: 9 - v[0] ** 2 - v[1] ** 2},))
_c10 = lambda: _solve(lambda v: np.sin(2 * v[0]) + np.cos(v[0]), [(-2, 3)], 100, "simplicial")
_c11 = lambda: _solve(lambda v: v[0] ** 2 + v[1] ** 2, [(-10, 10)] * 2, 100, "simplicial",
                      ({"type": "ineq", "fun": lambda v: v[0] + v[1] - 1},))

bench("01 Eggholder 2D", _c01)
check("01 Eggholder 2D", _val(_c01(), 1e0))

bench("02 Schwefel 2D", _c02)
check("02 Schwefel 2D", _val(_c02(), 1e3))

bench("03 Schaffer N2 2D", _c03)
check("03 Schaffer N2 2D", _val(_c03(), 1e4))

bench("04 Mishra Bird disk", _c04)
check("04 Mishra Bird disk", _val(_c04(), 1e3))

bench("05 Rugged sine 2D", _c05)
check("05 Rugged sine 2D", _val(_c05(), 1e4))

bench("06 Ackley 3D", _c06)
check("06 Ackley 3D", _val(_c06(), 1e4))

bench("07 Rastrigin 5D", _c07)
check("07 Rastrigin 5D", _val(_c07(), 1e4))

bench("08 Griewank 5D", _c08)
check("08 Griewank 5D", _val(_c08(), 1e4))

bench("09 Disk linear", _c09)
check("09 Disk linear", _val(_c09(), 1e4))

bench("10 Chained trig 1D", _c10)
check("10 Chained trig 1D", _val(_c10(), 1e4))

bench("11 Quadratic halfplane", _c11)
check("11 Quadratic halfplane", _val(_c11(), 1e4))
